src/main/cpp/src/map_zip_with_utils.cu

```diff
@@ -0,0 +1,368 @@
+ * limitations under the License.
+ */
+
+#include "map_zip_with_utils.hpp"
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/labeling/label_segments.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/lists/count_elements.hpp>
+#include <cudf/lists/gather.hpp>
+#include <cudf/lists/set_operations.hpp>
+#include <cudf/reduction.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/unary.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <thrust/scan.h>
+
+using namespace cudf;
+
+namespace spark_rapids_jni {
+
+// Taken from src/lists/utilities.cu
+std::unique_ptr<column> generate_labels(
+  lists_column_view const& input,
+  size_type n_elements,
+  rmm::cuda_stream_view stream      = cudf::get_default_stream(),
+  rmm::device_async_resource_ref mr = cudf::get_current_device_resource_ref())
+{
+  auto labels = make_numeric_column(
+    data_type(type_to_id<size_type>()), n_elements, cudf::mask_state::UNALLOCATED, stream, mr);
+  auto const labels_begin = labels->mutable_view().template begin<size_type>();
+  cudf::detail::label_segments(
+    input.offsets_begin(), input.offsets_end(), labels_begin, labels_begin + n_elements, stream);
+  return labels;
+}
+
+/*
+ *
+ * Example using two list columns and indexing
+ *
+ * search_keys     | search_values
+ * [1, 2, 3]       | [10, 2, 3]
+ * [4, 5]          | [4, 14, 5, 16]
+ * [6, 7, 8, 9]    | [6, 8]
+
+ * search_keys child:      1, 2, 3, 4,  5,  6, 7, 8, 9
+ * search_values child:   10, 2, 3, 4, 14, 16, 6, 8
+ *
+ * This algorithm works by comparing each key and value parallelly,
+ * hence the need for separate indexes. The algorithm first calculates
+ * the total number of comparisons need, which is a row-wise dot product.
+ * Then for each comparison, it generates the key index and value index which
+ * is used to compare the associated key and value.
+ *
+ * keys_labels:            0, 0, 0, 1,  1,  2, 2, 2, 2
+ * search_key_to_num_search_values: 3, 3, 3, 4, 4, 2, 2, 2, 2
+ * list_sizes_offsets: 0, 3, 6, 9, 13, 17, 19, 21, 23, 25
+ * key_index:   0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8
+ *
+ * total_compares_per_row: 9, 8, 8
+ * total_compares_offsets: 0, 9, 17, 25
+ * val_offsets: 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2
+ * values_idx:  0, 1, 2, 0, 1, 2, 0, 1, 2, 3, 4, 5, 6, 3, 4, 5, 6, 7, 8, 7, 8, 7, 8, 7, 8
+*/
+std::unique_ptr<column> indices_of(
+  lists_column_view const& search_keys,  // Column containing lists of keys to search for
+  lists_column_view const&
+    search_values,  // Column containing lists of values to search through (i.e. search space)
+  rmm::cuda_stream_view stream =
+    cudf::get_default_stream(),               // CUDA stream for asynchronous execution
+  rmm::device_async_resource_ref mr =
+    cudf::get_current_device_resource_ref())  // Memory resource for allocations
```

Comment:
I'm not sure `clang-format` has run on this.  The formatting looks completely off.